safeeyes/model.py

```diff
@@ -115,82 +125,86 @@ def is_long_break(self):
         return self.__current_break is not None and self.__current_break.type == BreakType.LONG_BREAK
 
     def next(self):
+        break_obj = None
+        shorts = self.__short_queue
+        longs  = self.__long_queue
+
         if self.is_empty():
             return None
-        break_obj = None
-        if self.__short_pointer is None:
-            # No short breaks
-            break_obj = self.__long_pointer
-            self.context['break_type'] = 'long'
-            # Update the pointer to next
-            self.__long_pointer = self.__long_pointer.next
-        elif self.__long_pointer is None:
-            # No long breaks
-            break_obj = self.__short_pointer
-            self.context['break_type'] = 'short'
-            # Update the pointer to next
-            self.__short_pointer = self.__short_pointer.next
-        elif self.__long_pointer.time <= self.__short_pointer.time:
-            # Time for a long break
-            break_obj = self.__long_pointer
-            self.context['break_type'] = 'long'
-            # Update the pointer to next
-            self.__long_pointer = self.__long_pointer.next
+
+        if shorts is None:
+            break_obj = self.__next_long()
+        elif longs is None:
+            break_obj = self.__next_short(ignore_time=True)
+        elif longs[self.__current_long].time <= shorts[self.__current_short].time:
+            break_obj = self.__next_long()
         else:
-            # Time for a short break
-            break_obj = self.__short_pointer
-            self.context['break_type'] = 'short'
-            # Reduce the break time from the next long break
-            self.__long_pointer.time -= self.__short_pointer.time
-            # Update the pointer to next
-            self.__short_pointer = self.__short_pointer.next
-
-        if self.__first_break is None:
-            self.__first_break = break_obj
-        self.context['new_cycle'] = self.__first_break == break_obj
-        if self.__current_break is not None:
-            # Reset the time of long breaks
-            if self.__current_break.type == BreakType.LONG_BREAK:
-                self.__current_break.time = self.__long_break_time
+            break_obj = self.__next_short()
+
+        if self.__shorts_taken == len(self.__short_queue):
```

Comment:
Is it enough to check the no of short breaks? Don't we need to consider the long breaks?

Suppose there are two short breaks (at 10 min interval) and a long break (at 20 min interval), cycle ends with a long break. Also consider no short breaks at all/no long breaks at all scenarios.

Reply:
I didn't change the default behavior for normal scenario.

In older master branch at line 150:
`self.context['new_cycle'] = self.__first_break == break_obj`

Because __first_break is normally a short break, in previos SafeEyes versions 'new_cycle' was set true every time we looped through the short breaks ie every time "Tightly close your eyes" break happened. 

So in this old behavior cycle time depended on the number of breaks the user added.
In my commit 'new_cycle' is also set every time we loop through all short breaks.

Should I will write a commit which sets 'new_cycle' every time we spend enough time to trigger interval between to long breaks Option? It represents physiological meaning of a cycle well.

Reply:
In the current logic the `__first_break` can be either a short break or long break https://github.com/slgobinath/SafeEyes/blob/master/safeeyes/model.py#L148 so even if we don't have short breaks, it would work. With the new change, if we don't have short breaks, it will treat every break as a new cycle.

Here a cycle is defined as completing all breaks (all short and long breaks) so I think a logic like this should work:
```
if self.__shorts_taken == len(self.__short_queue) and self.__longs_taken == len(self.__long_queue):
```
WDYT?

Reply:
This one won't work becauese it sets a cycle only when short breaks and long breaks reset at the same time.

I think the best defenition of the cycle is the interval between two long breaks. Because user can configure it with existing settings and because it does not depend on the number  of breaks.

![image](https://user-images.githubusercontent.com/25302233/106086393-a5b23980-6154-11eb-9e65-e09ab4a26c58.png)

Although if you define a cycle as showing user all possible breaks, I can add flags __shorts_completed and __longs_completed and set a cycle when both flags are True.

Reply:
Done. Check this commit https://github.com/slgobinath/SafeEyes/pull/418/commits/708bd3cb7a5c4c7d906df311cf3f4ef00d860e10

I defined cycle as the long break interval.
No short breaks -> trigger cycle every long break
Short and long breaks -> trigger cycle every long break
No long breaks -> trigger cycle when enough short break happen to fill the interval